src/app/core/components/manage-tags/manage-tags.component.spec.ts

```diff
@@ -0,0 +1,159 @@
+  let defaultTags: TagVO[] = [];
+  async function defaultRender(tags: TagVO[] = defaultTags) {
+    return await shallow.render(`<pr-manage-tags [tags]="tags"></pr-manage-tags>`, {
+      bind: {
+        tags,
+      }
+    });
+  }
+  beforeEach(() => {
+    throwError = false;
+    deleted = false;
+    renamed = false;
+    deletedTag = null;
+    renamedTag = null;
+    defaultTags = [
+      new TagVO({
+        name: 'Tomato',
+        tagId: 2,
+      }),
+      new TagVO({
+        name: 'Potato',
+        tagId: 1,
+      }),
+    ];
+    shallow = new Shallow(ManageTagsComponent, DummyModule).mock(ApiService, mockApiService);
+  });
+
+  it('should exist', async () => {
+    const { element } = await shallow.render();
+    expect(element).not.toBeNull();
+  });
+
+  it('should have a sorted list of tags', async () => {
+    const { find, element } = await defaultRender();
+    expect(find('.tag').length).toBe(2);
+    expect(find('.tag')[0].nativeElement.textContent).toContain('Potato');
+  });
+
+  it('should have a delete button for each tag', async () => {
+    const { find, fixture, element, outputs } = await defaultRender();
+    expect(find('.delete').length).toBeGreaterThan(0);
+    expect(outputs.refreshTags.emit).not.toHaveBeenCalled();
+    find('.delete')[0].nativeElement.click();
+    await fixture.whenStable();
+    expect(deleted).toBeTruthy();
+    expect(deletedTag.name).toBe('Potato');
+    expect(outputs.refreshTags.emit).toHaveBeenCalled();
+    await fixture.detectChanges();
+    expect(find('.tag').length).toBe(1);
+  });
+
+  it('should be able to rename tags', async () => {
+    const { find, fixture, element, outputs } = await defaultRender();
+    expect(find('.edit').length).toBeGreaterThan(0);
+    find('.edit')[0].nativeElement.click();
+    await fixture.detectChanges();
+    expect(find('.tag input').length).toBe(1);
+    expect(find('.tag input').nativeElement.value).toBe('Potato');
+    find('.tag input').nativeElement.focus();
+    find('.tag input').nativeElement.value = 'Starchy Tuber';
+    find('.tag input').nativeElement.dispatchEvent(new Event('change'));
+    find('.tag input').nativeElement.form.dispatchEvent(new Event('submit'));
+    await fixture.whenStable();
+    await fixture.detectChanges();
+    expect(find('.tag input').length).toBe(0);
+    expect(renamed).toBeTruthy();
+    expect(renamedTag.name).toBe('Starchy Tuber');
+    expect(outputs.refreshTags.emit).toHaveBeenCalled();
+  });
+
+  it('can cancel out of renaming a tag', async () => {
+    const { find, fixture } = await defaultRender();
+    expect(find('.edit').length).toBeGreaterThan(0);
+    find('.edit')[0].nativeElement.click();
+    await fixture.detectChanges();
+    expect(find('.cancel').length).toBe(1);
+    find('.cancel').nativeElement.click();
+    await fixture.detectChanges();
+    expect(find('.cancel').length).toBe(0);
+  });
+
+  it('should be able to filter the tags list', async () => {
```

Comment:
Let's also break this up into multiple tests. It looks like each instance of this is the same, with only the input and expected count differing; I believe in Jest that can most easily be done with the [`each` method](https://jestjs.io/docs/api#testeachtablename-fn-timeout).